### PR TITLE
Add MAGE_RUN_TYPE+MAGE_RUN_CODE to FPC identifier

### PR DESCRIPTION
**Pull request by amenk**
### Description

Currently Magento 2.2.6 full page cache does not work correctly, when using SetEnv or similar techniques on unchanged URLs

So the MAGE_RUN_TYPE and MAGE_RUN_CODE should be added to the FPC identifier to allow for such use cases.

A common case is Magento 2 behind a reverse NGINX proxy.

### Manual testing scenarios

Steps to reproduce, without proxy setup:
1) In pub/.htaccess set

SetEnv MAGE_RUN_TYPE store SetEnv MAGE_RUN_CODE italy

2) Now open Frontend

3) Now set

SetEnv MAGE_RUN_TYPE store SetEnv MAGE_RUN_CODE english

4) Now Open frontend, same URL

-> Italian version is displayed

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18624) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @amenk. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by miguelbalparda** [comment]
Hi @amenk, can you please check the failing tests?

**Comment by amenk** [comment]
@miguelbalparda

Okay I will..probably I should also add a new test as well.
Is the code fine, or is the approach wrong?

I was afraid using mage store constants in framework but it seems to happen also in other places.

**Comment by miguelbalparda** [comment]
I think the approach is correct, will double check internally and get back to you if anything changes.

**Comment by miguelbalparda** [comment]
Anything on this one @amenk? It's been a while, please let me know if you are still working on this PR.

**Comment by amenk** [comment]
> I think the approach is correct, will double check internally and get back to you if anything changes.

@miguelbalparda I was waiting for your response on this ...

and I am having a lack of time, but will try my best to get that PR through, if you tell me, the approach is correct.

**Comment by miguelbalparda** [comment]
Closing for lack of feedback, feel free to reopen if you have time to keep working on this,

**Comment by amenk** [comment]
@miguelbalparda 

>  will double check internally and get back to you if anything changes.

Did you already double check?

**Comment by amenk** [comment]
@miguelbalparda your feedback would motivate me :-) 

I am still not sure if my approach has scoping issues ...

**Comment by amenk** [comment]
Will work on this today.

**Review by phoenix128 — APPROVED** [review]
Looks good to me.
Thank you for your time and your contribution.

**Comment by magento-engcom-team** [comment]
Hi @phoenix128, thank you for the review.
[ENGCOM-4467](https://jira.corp.magento.com/browse/ENGCOM-4467) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @amenk , thanks for the pull request!

I have a suggestion regarding the adjustment of the fix.

From my point of view, `MAGE_RUN_TYPE` and `MAGE_RUN_CODE` parameters are used as just a method to set the current store view. Potentially there can be other ways to set the current store view. And that's actually a store view that is responsible for differentiating content.

I think it would be better to ensure that page cache is unique per store view rather than using input parameters as cache identifiers. That can be done by adding current store view code to the FPC identifier.

(In case `website` is used as `MAGE_RUN_TYPE` - the identifier of default store view for the requested website should be used for cache variation)

Please correct me if I'm wrong or missing anything and let me know what do you think about my suggestion.

**Comment by amenk** [comment]
@sivaschenko 
Thank you for your review and suggestions. It sounds right, I did not think about that.

I am just not sure if the store view is available at the early stages of FPC.

Do you want to improve the PR or shall I look into it when I find a time?

**Comment by sivaschenko** [comment]
@amenk yes, cache identifier should be based on store view.

Actually, as I can see store view is added to `Magento\Framework\App\Http\Context` in `\Magento\Store\App\Action\Plugin\Context::beforeDispatch` and the http context is used to generate FPC identifier in `\Magento\Framework\App\PageCache\Identifier::getValue` (`\Magento\PageCache\Model\App\FrontController\BuiltinPlugin::aroundDispatch`)

Probably the bug appears because of the order of these plugins

**Comment by amenk** [comment]
Wow that is more than interesting. Wondering if that is only on since 2.3.x
or already in 2.2.x where we faced the problem. So my patch might not be
needed at all.

Have to check...

**Comment by amenk** [comment]
@sivaschenko  I think it cannot work as suggested. If I put a 

    die(__METHOD__);

into `\Magento\Store\App\Action\Plugin\Context::beforeDispatch` (after the current page was cached) and refresh the page, this is not called at all (also not too late) if the file is in he full page cache.

But the information about the run code has to be available on an early stage. During the bootstrapping of FPC the above method nevers seems to be run.

**Comment by amenk** [comment]
@sivaschenko I believe my code is needed :-) What do you think?

**Comment by amenk** [comment]
@sivaschenko What do you think?

**Comment by sivaschenko** [comment]
@amenk thanks for your patience, this pull request is under discussion

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by amenk** [comment]
> @amenk unfortunately, only members of the maintainers team are allowed to unassign developers from the pull request

I did not try to unassign anybody :)

**Comment by amenk** [comment]
all checks are green now

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4467](https://jira.corp.magento.com/browse/ENGCOM-4467) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Pull request merged**
